agent/lib_aws_sdk_php.c

```diff
@@ -14,10 +14,409 @@
+ * AWS Service, gets the guzzle response from the AWS Service, and bundles that
+ * response into an AswResult to return.
+ *
+ * @param string $name      Name of the command to execute.
+ * @param array  $arguments Arguments to pass to the getCommand method.
+ *
+ * @return ResultInterface
+ * @throws \Exception
+ */
+
+NR_PHP_WRAPPER(nr_aws_client_call_before) {
+  (void)wraprec;
+  nr_segment_t* segment = NULL;
+  /*
+   * Start a segment in case it needs to become an external, message, or
+   * datastore segment.
+   */
+  segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+  if (NULL != segment) {
+    segment->wraprec = auto_segment->wraprec;
+  }
+}
+NR_PHP_WRAPPER_END
+
+NR_PHP_WRAPPER(nr_aws_client_call) {
+  (void)wraprec;
+
+  zval* command_name = NULL;
+  const char* klass = NULL;
+  char* command_name_string = NULL;
+  char* real_class_and_command = NULL;
+  nr_segment_t* segment = NULL;
+  zend_class_entry* class_entry = NULL;
+  int klass_len = 0;
+
+  class_entry = Z_OBJCE_P(nr_php_execute_scope(execute_data));
+  klass = nr_php_class_entry_name(class_entry);
+
+  if (NULL != klass) {
+    /* Get the arg command_name. */
+    command_name = nr_php_arg_get(1, NR_EXECUTE_ORIG_ARGS);
+
+    if (nr_php_is_zval_non_empty_string(command_name)) {
+      command_name_string = Z_STRVAL_P(command_name);
+      klass_len = nr_php_class_entry_name_length(class_entry);
+
+      if (klass_len == AWS_SDK_PHP_SQSCLIENT_CLASS_LEN
+          && nr_striendswith(klass, klass_len,
+                             AWS_SDK_PHP_SQSCLIENT_CLASS_SHORT,
+                             AWS_SDK_PHP_SQSCLIENT_CLASS_SHORT_LEN)) {
+        nr_lib_aws_sdk_php_sqs_handle(auto_segment, command_name_string,
+                                      Z_STRLEN_P(command_name),
+                                      NR_EXECUTE_ORIG_ARGS);
+      }
+
+      if (NR_SEGMENT_CUSTOM == auto_segment->type) {
+        /*
+         * We need to end the segment that we started in the 'before' wrapper if
+         * it wasn't handled and ended by the handling function. Handling
+         * function would have changed the segment type from from default
+         * (`NR_SEGMENT_CUSTOM`) if it ended it.
+         */
+        nr_segment_discard(&auto_segment);
+      }
+
+      /*
+       * Since we have klass and command_name, we can give the calling segment
+       * a more meaningful name than Aws/AwsClient::__call We can decode it to
+       * Aws/CALLING_CLASS_NAME/CALLING_CLASS_CLIENT::CALLING_CLASS_COMMAND
+       *
+       * EX: Aws\\Sqs\\SqsClient::sendMessage
+       */
+
+      segment = nr_txn_get_current_segment(NRPRG(txn), NULL);
+      if (NULL != segment) {
+        real_class_and_command
+            = nr_formatf("Custom/%s::%s", klass, command_name_string);
+        nr_segment_set_name(segment, real_class_and_command);
+        nr_free(real_class_and_command);
+      }
+    }
+    /* Release the command_name. */
+    nr_php_arg_release(&command_name);
+  }
```

Comment:
newrelic-php-agent's code base doesn't seem to have a standard around structuring the code. Golang has the following standard defined in [Effective Go](https://go.dev/doc/effective_go#if):
> [...] a common situation where code must guard against a sequence of error conditions. The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise. Since error cases tend to end in return statements, the resulting code needs no else statements.

The readability of the above could could definitely benefit from applying Effective Go standards. Consider:
```
  if (NULL == klass) {
    return;
  }
  ...
  if (!nr_php_is_zval_non_empty_string(command_name)) {
    nr_php_arg_release(&command_name);
    return;
  }
  
```

Reply:
This suggestion is something to consider but is not blocking approval.

Reply:
`nr_lib_aws_sdk_php_sqs_parse_queueurl` follows the Effective Go standards and is easy to read.

Reply:
Agree, but it's not a simple return that can occur since it needs a return value.
I can however, make a `goto end` so that we'll skip to the bottom then all the [logic here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_wrapper.h#L274) can execute for the actual return value.

Reply:
See the link in the previous comment to see how the return value is computed on wrappers.

Reply:
This is what is needed to "exit" early from a wrapper: [1523be8](https://github.com/newrelic/newrelic-php-agent/pull/1003/commits/1523be8657cb7c18c98f51cad3e65250005a2063)

Reply:
I see. Also in case there's no custom logic is needed (like freeing memory), instead of a simple `return`, [`NR_PHP_WRAPPER_LEAVE`](https://github.com/newrelic/newrelic-php-agent/blob/R11.5/agent/php_wrapper.h#L305C9-L305C29) can also be used, which will `goto` [here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_wrapper.h#L275) and end the wrapper correctly.